hypothesis-python/src/hypothesis/strategies/_internal/numbers.py

```diff
@@ -416,10 +449,38 @@ def floats(
     elif max_value == -math.inf:
         raise InvalidArgument("allow_infinity=False excludes max_value=-inf")
 
+    smallest_normal = width_smallest_normals[width]
+    if allow_subnormal is None:
+        if min_value is not None and max_value is not None:
+            if min_value == max_value:
+                allow_subnormal = -smallest_normal < min_value < smallest_normal
+            else:
+                allow_subnormal = (
+                    min_value < smallest_normal and max_value > -smallest_normal
+                )
+        elif min_value is not None:
+            allow_subnormal = min_value < smallest_normal
+        elif max_value is not None:
+            allow_subnormal = max_value > -smallest_normal
+        else:
+            allow_subnormal = True
+    if allow_subnormal and min_value is not None and min_value >= smallest_normal:
+        raise InvalidArgument(
+            f"allow_subnormal=True, but minimum value {min_value} "
+            f"excludes values below smallest positive normal {smallest_normal}"
+        )
+    if allow_subnormal and max_value is not None and max_value <= -smallest_normal:
+        raise InvalidArgument(
+            f"allow_subnormal=True, but maximum value {max_value} "
+            f"excludes values above smallest negative normal {-smallest_normal}"
+        )
+
     unbounded_floats = FloatStrategy(
-        allow_infinity=allow_infinity, allow_nan=allow_nan, width=width
+        allow_infinity=allow_infinity,
+        allow_nan=allow_nan,
+        allow_subnormal=allow_subnormal,
+        width=width,
     )
```

Comment:
Given how often we repeat these kwargs, I think I'd prefer to read them as

```python
    kw = {"allow_subnormal": allow_subnormal, "width": width}
    unbounded_floats = FloatStrategy(
        allow_infinity=allow_infinity, allow_nan=allow_nan, **kw
    )
```

and so on.  It should at least make the diff a bit smaller!

Reply:
Ah so I did this, but MyPy didn't like it. Maybe type hinting kw with something could work, hmm.

Reply:
Ah, I'd just silence it with `kw: Any = {...}` or maybe `# type: ignore`  - we annotate the public API so that users get better autocomplete and linting, but our internals are often inexpressible.

(that's *improving* with the steady stream of typing PEPs like `ParamSpec`, but the gap remains)